staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -147,14 +153,41 @@ type Authenticator struct {
 	// The mutex also guards calling the plugin. Since the plugin could be
 	// interactive we want to make sure it's only called once.
 	mu          sync.Mutex
-	cachedToken string
+	cachedCreds *credentials
 	exp         time.Time
+
+	onRotate func()
 }
 
-// WrapTransport instruments an existing http.RoundTripper with credentials returned
-// by the plugin.
-func (a *Authenticator) WrapTransport(rt http.RoundTripper) http.RoundTripper {
-	return &roundTripper{a, rt}
+type credentials struct {
+	token string
+	cert  *tls.Certificate
+}
+
+// UpdateTransportConfig updates the transport.Config to use credentials
+// returned by the plugin.
+func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
+	wt := c.WrapTransport
+	c.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
+		if wt != nil {
+			rt = wt(rt)
+		}
+		return &roundTripper{a, rt}
+	}
+
+	c.TLS.GetCert = a.cert
+
+	defaultDialer := &net.Dialer{Timeout: 30 * time.Second, KeepAlive: 30 * time.Second}
+	ct := conntrack.NewTracker(func(ctx context.Context, network, addr string) (net.Conn, error) {
+		if c.Dial != nil {
```

Comment:
I think this is your bug. You don't save the dialer before setting it later like you do with the WrapTransport.
```
dial := c.Dial

ct := conntrack.NewTracker(.... {
    if dial != nil {
        return dial(network, addr)
    }
    return defaultDialer.DialContext(ctx, network, addr)
})
```

Reply:
Yep, you're right. That was it.